### PR TITLE
Update tool update support

### PR DESCRIPTION
**Pull request by warunalakshitha**
## Purpose
$Subject

### CODE REVIEW

File: resources/tools/scripts/install [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    CURRENT_PATH="$(dirname "$(readlink -f "$0")")"
+fi
+
+\cp $CURRENT_PATH/ballerina-command-${ballerina.command.version}/lib/ballerina-command-${ballerina.command.version}.jar $CURRENT_PATH/../lib
+\cp $CURRENT_PATH/ballerina-command-${ballerina.command.version}/bin/ballerina $CURRENT_PATH/../bin
+for file in $CURRENT_PATH/../lib/*; do
+    if [ -f "$file" ] && [[ "$file" == *ballerina-command*.jar ]] && [[ "$file" != *ballerina-command-${ballerina.command.version}.jar ]]; then
+	rm -rf $file
+    fi
+done
```
Comment (Kishanthan) [inline resources/tools/scripts/install]: new line missing

File: ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public String getName() {
+        return BallerinaCliCommands.UPDATE;
+    }
+
+    @Override
+    public void printLongDesc(StringBuilder out) {
+
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java]: Not providing long desc now?

File: ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    public void printUsage(StringBuilder out) {
+
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java]: remove this extra new line

File: ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java [hunk 4]
```diff
@@ -207,6 +210,38 @@ public static String getValue(String key, String json) {
...
+        sc.init(null, trustAllCerts, new java.security.SecureRandom());
+        HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
+
+        URL url = new URL(PRODUCTION_URL + "/versions/latest");
+
+        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+        conn.setRequestMethod("GET");
+        conn.setRequestProperty("user-agent",
+                                OSUtils.getUserAgent(getCurrentBallerinaVersion(),
+                                                     getCurrentToolsVersion(), "jballerina"));
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java]: check whether these three params are aligned correctly.

File: ballerina-command/src/main/java/org/ballerinalang/command/util/OSUtils.java [hunk 5]
```diff
@@ -67,6 +67,21 @@ public static String getExecutableFileName() {
...
+    /**
+     * Provide file name of tool update install script for current operating system.
+     * @return name of the file
+     */
+    public static String getInstallScriptFileName() {
+        if (OSUtils.isWindows()) {
+            return "install.bat";
+        } else if (OSUtils.isUnix() || OSUtils.isSolaris()) {
+            return "install";
+        } else if (OSUtils.isMac()) {
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/util/OSUtils.java]: this else if can be merged with above else if ?

File: ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java [hunk 6]
```diff
@@ -326,6 +377,66 @@ private static void downloadAndSetupDist(PrintStream printStream, HttpURLConnect
...
+            SSLContext sc = SSLContext.getInstance("SSL");
+            sc.init(null, trustAllCerts, new java.security.SecureRandom());
+            HttpsURLConnection.setDefaultSSLSocketFactory(sc.getSocketFactory());
+
+            URL url = new URL(ToolUtil.PRODUCTION_URL + "/versions/" + toolVersion);
+            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+            conn.setRequestMethod("GET");
+            conn.setRequestProperty("user-agent",
+                                    OSUtils.getUserAgent(getCurrentBallerinaVersion(),
+                                                         getCurrentToolsVersion(), "jballerina"));
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java]: these three params alignment too

File: ballerina-command/src/main/java/org/ballerinalang/command/util/OSUtils.java (hunk 5 above)
Comment (warunalakshitha) [reply ballerina-command/src/main/java/org/ballerinalang/command/util/OSUtils.java]: Implemented the same way as current getExecutableFileName method.

File: ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java [hunk 7]
```diff
@@ -326,6 +377,66 @@ private static void downloadAndSetupDist(PrintStream printStream, HttpURLConnect
...
+            if (conn.getResponseCode() != 200) {
+                throw new RuntimeException("Failed : HTTP error code : " + conn.getResponseCode());
+            }
+            conn.disconnect();
+        } else {
+            printStream.println("Current user does not have write permissions to directory");
+        }
+    }
+
+
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/util/ToolUtil.java]: unwanted extra new lines here

File: ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+import org.ballerinalang.command.util.ToolUtil;
+import picocli.CommandLine;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+
+/**
+ * This class represents the "Update" command and it holds arguments and flags specified by the user.
```
Comment (Kishanthan) [inline ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java]: New class missing @ since tag.

Comment (warunalakshitha) [reply ballerina-command/src/main/java/org/ballerinalang/command/cmd/UpdateToolCommand.java]: Seems all classes missed. Fixed for update command. We can others that in seperate pr.

**Pull request merged**